src/helpers/useFocusItemHelper.ts

```diff
@@ -0,0 +1,171 @@
+import _ from 'lodash';
+import {RefObject, useState, useCallback, useEffect, useRef} from 'react';
+import {LayoutChangeEvent, ScrollView, FlatList} from 'react-native';
+import {Constants} from '.';
+
+export enum OffsetType {
+  CENTER = 'CENTER',
+  DYNAMIC = 'DYNAMIC',
+  LEFT = 'LEFT',
+  RIGHT = 'RIGHT'
+}
+
+export type Props = {
+  /**
+   * A reference to the ScrollView (or FlatList) which the items are in
+   */
+  scrollViewRef: RefObject<ScrollView | FlatList>;
+  /**
+   * The number of items
+   */
+  itemsCount: number;
+  /**
+   * The selected item's index
+   */
+  selectedIndex?: number;
+  /**
+   * Where would the item be located (default to CENTER)
+   */
+  offsetType?: OffsetType;
+  /**
+   * Add a margin to the offset (default to true)
+   * This gives a better UX
+   * Not relevant to OffsetType.CENTER
+   */
+  addOffsetMargin?: boolean;
+  /**
+   * How much space (padding \ margin) is there on he left\right of the items
+   */
+  sideSpacing?: number;
+  /**
+   * How much space (padding \ margin) is there between each item
+   */
+  innerSpacing?: number;
+};
+
+// TODO: this is what I want, is there a better way to do it?
+type Offsets = {
+  CENTER: number[];
+  LEFT: number[];
+  RIGHT: number[];
+};
+
+export type ResultProps = {
+  /**
+   * This should be called by each ot the items' onLayout
+   */
+  onItemLayout: (event: LayoutChangeEvent, index: number) => void;
+  /**
+   * The items' width
+   */
+  itemsWidths: number[];
+  /**
+   * Use in order to focus the item with the specified index
+   */
+  focusIndex: (index: number, animated?: boolean) => void;
+};
+
+const useFocusItemsHelper = (props: Props): ResultProps => {
+  const {
+    scrollViewRef,
+    itemsCount,
+    selectedIndex,
+    offsetType = OffsetType.CENTER,
+    addOffsetMargin = true,
+    sideSpacing = 0,
+    innerSpacing = 0
+  } = props;
+  const itemsWidths = useRef<(number | null)[]>(_.times(itemsCount, () => null));
+  const currentIndex = useRef<number>(selectedIndex || 0);
+  const [offsets, setOffsets] = useState<Offsets>({CENTER: [], LEFT: [], RIGHT: []});
```

Comment:
Do we need to keep the offsets in state?
They don't **really** affect the render..

Reply:
Actually it does (verified it now), `focusIndex` needs to change when the `offsets` have been calculated.